agent/exec/controller.go

```diff
@@ -131,23 +141,30 @@ func Do(ctx context.Context, task *api.Task, ctlr Controller) (*api.TaskStatus,
 			panic("err must not be nil when fatal")
 		}
 
-		if IsTemporary(err) {
-			switch Cause(err) {
-			case context.DeadlineExceeded, context.Canceled:
-				// no need to set these errors, since these will more common.
-			default:
-				status.Err = err.Error()
+		if cs, ok := err.(ContainerStatuser); ok {
+			var err error
+			containerStatus, err = cs.ContainerStatus(ctx)
+			if err != nil {
+				log.G(ctx).WithError(err).Error("error resolving container status on fatal")
 			}
+		}
 
+		// make sure we've set the *correct* exit code
+		if ec, ok := err.(ExitCoder); ok {
+			exitCode = ec.ExitCode()
+		}
+
+		if cause := errors.Cause(err); cause == context.DeadlineExceeded || cause == context.Canceled {
 			return retry()
 		}
 
-		if cause := Cause(err); cause == context.DeadlineExceeded || cause == context.Canceled {
+		status.Err = err.Error() // still reported on temporary
+		if IsTemporary(err) {
 			return retry()
 		}
 
+		// only at this point do we consider the error fatal to the task.
 		log.G(ctx).WithError(err).Error("fatal task error")
-		status.Err = err.Error()
 
 		switch {
 		case status.State < api.TaskStateStarting:
```

Comment:
Although if it fails to start it's likely that it will never start again. It's probably a configuration error that can only be discovered on runtime. In contrast, if something fails from a running state it will probably start fine next time.